File: src/agents/planner.js
import { retrieveIssueContext } from "../graph/retriever.js";

export class PlannerAgent {
  constructor({ model, logger } = {}) {
    this.model = model;
    this.logger = logger;
  }

  async run(state) {
    const context = await retrieveIssueContext({
      graph: state.graph,
      repoPath: state.repoPath,
      issue: state.issue,
      intent: state.intent
    });

    if (this.model) {
      const plan = await this.fromModel(state, context);
      return { ...plan, context };
    }

    return heuristicPlan(state, context);
  }

  async fromModel(state, context) {
    const prompt = `Plan a small production-quality fix for this Go repository issue.

Repository: ${state.repoFullName}
Issue: #${state.issue.number} ${state.issue.title}
Intent:
${JSON.stringify(state.intent, null, 2)}

Graph-ranked context:
${JSON.stringify(context.graphNodes.slice(0, 20), null, 2)}

File excerpts:
${context.snippets.map((snippet) => `--- ${snippet.path}\n${snippet.excerpt}`).join("\n\n")}

Return strict JSON:
{
  "summary": "one sentence",
  "filesToEdit": ["relative/path.go"],
  "testsToEdit": ["relative/path_test.go"],
  "targetedCommands": ["go test -run TestName ./..."],
  "standardCommands": ["go test ./..."],
  "patchBudget": {"maxFiles": 3, "maxLinesChanged": 120},
  "risks": ["risk"],
  "reviewChecklist": ["check"]
}`;

    return await this.model.generateJson({
      system: "You are a conservative Go maintainer. Prefer minimal behavioral patches and focused tests.",
      prompt,
      temperature: 0.15
    });
  }
}

export function heuristicPlan(state, context) {
  const project = state.project;
  const files = new Set([
    ...(state.intent.candidateFiles ?? []),
    ...context.files.slice(0, 4).map((file) => file.path)
  ]);

  const tests = [...files].filter((file) => file.endsWith("_test.go"));
  if (!tests.length) {
    const likelyTest = context.files.find((file) => file.path.endsWith("_test.go"))?.path;
    if (likelyTest) {
      tests.push(likelyTest);
    } else if (state.repoFullName === "go-playground/validator") {
      tests.push("validator_test.go");
    }
  }

  return {
    summary: `Fix ${state.issue.title} with a small code change and focused tests.`,
    filesToEdit: [...files].filter((file) => !file.endsWith("_test.go")).slice(0, 4),
    testsToEdit: tests.slice(0, 3),
    targetedCommands: inferTargetedCommands(state),
    standardCommands: project.standardChecks,
    patchBudget: {
      maxFiles: 3,
      maxLinesChanged: 120
    },
    risks: [
      ...(project.riskNotes ?? []),
      ...state.intent.riskNotes
    ],
    reviewChecklist: [
      "Patch is limited to issue-related behavior.",
      "Tests cover the reported case and a nearby valid case.",
      "No exported API changes unless the issue requires them.",
      "No avoidable work added to hot paths."
    ],
    context
  };
}

function inferTargetedCommands(state) {
  const testSymbols = (state.intent.symbols ?? [])
    .filter((symbol) => /^test/i.test(symbol))
    .map((symbol) => symbol.replace(/[^A-Za-z0-9_]/g, ""));
  if (testSymbols.length) {
    return [`go test -run ${testSymbols[0]} ./...`];
  }

  const titleWords = state.issue.title
    .split(/[^A-Za-z0-9]+/)
    .filter((word) => word.length >= 5)
    .slice(0, 3)
    .join("|");
  return titleWords ? [`go test -run '${titleWords}' ./...`] : ["go test ./..."];
}

File: go_issue_agent/agents/planner.py
from __future__ import annotations

from go_issue_agent.graph.retriever import retrieve_issue_context


class PlannerAgent:
    def __init__(self, model=None, logger=None):
        self.model = model
        self.logger = logger

    def run(self, state: dict) -> dict:
        context = retrieve_issue_context(
            graph=state["graph"],
            repo_path=state["repoPath"],
            issue=state["issue"],
            intent=state["intent"],
        )
        if self.model:
            plan = self.from_model(state, context)
            plan["context"] = context
            return plan
        return heuristic_plan(state, context)

    def from_model(self, state: dict, context: dict) -> dict:
        snippets_text = "".join(
            f"--- {snippet['path']}\n{snippet['excerpt']}\n\n"
            for snippet in context["snippets"]
        )
        prompt = f"""Plan a small production-quality fix for this Go repository issue.

Repository: {state["repoFullName"]}
Issue: #{state["issue"]["number"]} {state["issue"]["title"]}
Intent:
{state["intent"]}

Graph-ranked context:
{context["graphNodes"][:20]}

File excerpts:
{snippets_text}

Return strict JSON:
{{
  "summary": "one sentence",
  "filesToEdit": ["relative/path.go"],
  "testsToEdit": ["relative/path_test.go"],
  "targetedCommands": ["go test -run TestName ./..."],
  "standardCommands": ["go test ./..."],
  "patchBudget": {{"maxFiles": 3, "maxLinesChanged": 120}},
  "risks": ["risk"],
  "reviewChecklist": ["check"]
}}"""
        return self.model.generate_json(
            system="You are a conservative Go maintainer. Prefer minimal behavioral patches and focused tests.",
            prompt=prompt,
            temperature=0.15,
        )


def heuristic_plan(state: dict, context: dict) -> dict:
    project = state["project"]
    files = list(dict.fromkeys([*(state["intent"].get("candidateFiles") or []), *[file["path"] for file in context["files"][:4]]]))
    tests = [file for file in files if file.endswith("_test.go")]
    if not tests:
        likely_test = next((file["path"] for file in context["files"] if file["path"].endswith("_test.go")), None)
        if likely_test:
            tests.append(likely_test)
        elif state["repoFullName"] == "go-playground/validator":
            tests.append("validator_test.go")
    return {
        "summary": f'Fix {state["issue"]["title"]} with a small code change and focused tests.',
        "filesToEdit": [file for file in files if not file.endswith("_test.go")][:4],
        "testsToEdit": tests[:3],
        "targetedCommands": infer_targeted_commands(state),
        "standardCommands": project["standardChecks"],
        "patchBudget": {"maxFiles": 3, "maxLinesChanged": 120},
        "risks": [*(project.get("riskNotes") or []), *(state["intent"].get("riskNotes") or [])],
        "reviewChecklist": [
            "Patch is limited to issue-related behavior.",
            "Tests cover the reported case and a nearby valid case.",
            "No exported API changes unless the issue requires them.",
            "No avoidable work added to hot paths.",
        ],
        "context": context,
    }


def infer_targeted_commands(state: dict) -> list[str]:
    test_symbols = []
    for symbol in state["intent"].get("symbols") or []:
        if symbol.lower().startswith("test"):
            cleaned = "".join(ch for ch in symbol if ch.isalnum() or ch == "_")
            test_symbols.append(cleaned)
    if test_symbols:
        return [f"go test -run {test_symbols[0]} ./..."]

    title_words = [word for word in "".join(ch if ch.isalnum() else " " for ch in state["issue"]["title"]).split() if len(word) >= 5][:3]
    return [f"go test -run '{'|'.join(title_words)}' ./..."] if title_words else ["go test ./..."]
